app/models/service_terraform_template.rb

```diff
@@ -55,15 +73,34 @@ def check_completed(action)
   end
 
   def launch_terraform_template(action)
+    $embedded_terraform_log.debug("Service(#{id}).launch_terraform_template(#{action}) starts")
     terraform_template = terraform_template(action)
 
-    # runs provision or retirement job, based on job_options
+    # runs provision or retirement or reconfigure job, based on job_options
     stack = ManageIQ::Providers::EmbeddedTerraform::AutomationManager::Stack.create_stack(terraform_template, get_job_options(action))
+
+    # We save the newly created Job/OrchestrationStack.id, with job options, to identify the current running Job/OrchestrationStack.
+    # This especially is needed for Reconfigure action, as the action can run multiple times for same service instance, which creates multiple OrchestrationStack(job)s,
+    # unlike Provision or Retirement actions, each having single OrchestrationStack(job).
+    save_orchestration_stack_id(action, stack.id)
+
     add_resource!(stack, :name => action)
+
+    $embedded_terraform_log.debug("Service(#{id}).launch_terraform_template(#{action}) ends")
   end
 
   def stack(action)
-    service_resources.find_by(:name => action, :resource_type => 'OrchestrationStack').try(:resource)
+    job_options = options[job_option_key(action)]
+    unless job_options && job_options.key?(:orchestration_stack_id)
```

Comment:
I avoid using `unless` with anything more complicated than a single conditional.  Anything that tacks on `&&` or `||` and it takes up more brain compute cycles than is worth IMO.  Just use a normal `if job_options.nil? || !job_options.key?(:orchestration_stack_id)` or even better `if !job_options&.key?(:orchestration_stack_id)`